### PR TITLE
Add parser for southbound operations

### PR DESCRIPTION
**Pull request by alla3**
Add parser for the following messages:
- Data Update
- Request Date
- Schema Status
- Device Auth Status
- Registered Device
- Device Removed

Depends on #13

### CODE REVIEW

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageDataItem.kt [hunk 1]
```diff
@@ -6,5 +6,6 @@ package com.cesar.knot_sdk.knot_messages
  * @property data the data batch that is to be sent
  */
 data class KNoTMessageDataItem(
+    val sensorId : Int,
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageDataItem.kt]: As we discussed, this is not necessary, please remove it.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 2]
```diff
@@ -8,7 +8,9 @@ import com.cesar.knot_sdk.KNoTTypes.KNOT_VALUE_TYPE_INT
 import com.cesar.knot_sdk.KNoTTypes.KNOT_VALUE_TYPE_MAX
 import com.cesar.knot_sdk.KNoTTypes.KNOT_VALUE_TYPE_RAW
 import com.cesar.knot_sdk.knot_messages.KNoTMessageDataItem
+import com.cesar.knot_sdk.knot_messages.KNoTMessageRequestData
 import com.cesar.knot_sdk.knot_messages.KNoTMessageUpdateData
+
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: Why did you add this new line? If there is no particular reason, please remove it.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageAuthenticated.kt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package com.cesar.knot_sdk.knot_messages
+
+/**
+ *
+ */
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageAuthenticated.kt]: Why are you adding this comment section? If you are not going to make a commnet, please remove it.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageRegistered.kt [hunk 4]
```diff
@@ -7,5 +7,6 @@ package com.cesar.knot_sdk.knot_messages
  */
 data class KNoTMessageRegistered(
     val id : String,
-    val token : String
+    val token : String,
+    val error: String
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageRegistered.kt:12]: This looks different from the connector API, are you sure this is the expected message?

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * This class parses southbound traffic json requests in usable kotlin objects.
+ */
+class KNoTMessageParser {
+
+    private val sensorHash = mutableMapOf<Int, Int>()
+    private val VALUE = "value"
+    private val DATA = "data"
+    private val ID = "id"
+    private val SENSOR_ID = "sensorId"
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt:27]: Is there any reason not to use "sensorid" instead of "sensorId"?

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            "KNOT_VALUE_TYPE_INT, KNOT_VALUE_TYPE_FLOAT, " +
+            "KNOT_VALUE_TYPE_BOOL, " + "KNOT_VALUE_TYPE_RAW, KNOT_VALUE_NULL."
+
+    private fun getSensorType(sensorId : Int) : Int {
+        val sensorType = sensorHash[sensorId]
+
+        if(sensorType != null) {
+            return sensorType
+        } else {
+            Log.d("DEV-LOG", "This sensor has not been registered.")
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: Remove this log.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        return KNOT_VALUE_NULL
+    }
+
+    fun addSensor(key : Int, value : Int) {
+        if (value >= KNOT_VALUE_TYPE_MAX) {
+            throw IllegalArgumentException(
+                ILLEGAL_ARGUMENT_EXCEPTION_MESSAGE
+            )
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: What do you think about one-lining this statement?

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+                KNOT_VALUE_TYPE_FLOAT -> data.add(KNoTMessageDataItem(sensorId,
+                    it.asJsonObject.get(VALUE).asFloat))
+
+                KNOT_VALUE_TYPE_BOOL -> data.add(KNoTMessageDataItem(sensorId,
+                    it.asJsonObject.get(VALUE).asBoolean))
+
+                KNOT_VALUE_TYPE_RAW -> data.add(KNoTMessageDataItem(sensorId,
+                    it.asJsonObject.get(VALUE).asString))
+
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: This formatting looks weird, was this done by your IDE? I think it would look way better if you broke the line at the start f th add. 
e.g.: data.add(KNoTMessageDataItem(
                          sensorId,
                          it.asJsonObject.get(VALUE).asInt)
       )

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                KNOT_VALUE_NULL -> { }
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: This looks weird? When are you receiveng KNOT_VALUE_NULL exactly?

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 10]
```diff
@@ -78,4 +83,18 @@ class KNoTMessageParser {
 
         return KNoTMessageUpdateData(id, data)
     }
+
+    fun parseDataRequest(knotData : String) : KNoTMessageRequestData {
+        val jelement : JsonElement =  JsonParser.parseString(knotData)
+        val jobject = jelement.asJsonObject
+        val id = jobject.get(ID).asString
+        val jarray = jobject.getAsJsonArray(SENSOR_IDS)
+        val sensorIds = mutableListOf<Int>()
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: I understand that the naming of each variable is related to the object being returned by the right-hand-side of the assigment, but maybe you could change these variable names? jelement could be receivedJson for example and so on.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 11]
```diff
@@ -97,4 +98,12 @@ class KNoTMessageParser {
 
         return KNoTMessageRequestData(id, sensorIds)
     }
+
+    fun parseSchemaStatus(knotSchema : String) : KNoTMessageSchemaResp {
+        val jelement : JsonElement = JsonParser.parseString(knotSchema)
+        val id = jelement.asJsonObject.get(ID).asString
+        val error = jelement.asJsonObject.get(ERROR).toString()
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: See my previous comment about variable naming.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 12]
```diff
@@ -106,4 +107,12 @@ class KNoTMessageParser {
 
         return KNoTMessageSchemaResp(id, error)
     }
+
+    fun parseAuthStatus(knotAuth : String) : KNoTMessageAuthenticated {
+        val jelement : JsonElement = JsonParser.parseString(knotAuth)
+        val id = jelement.asJsonObject.get(ID).asString
+        val error = jelement.asJsonObject.get(ERROR).toString()
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: See my previous comment about variable naming.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 13]
```diff
@@ -115,4 +116,13 @@ class KNoTMessageParser {
 
         return KNoTMessageAuthenticated(id, error)
     }
+
+    fun parseDeviceRegistered(knotRegister : String) : KNoTMessageRegistered {
+        val jelement : JsonElement = JsonParser.parseString(knotRegister)
+        val id = jelement.asJsonObject.get(ID).asString
+        val token = jelement.asJsonObject.get(TOKEN).toString()
+        val error = jelement.asJsonObject.get(ERROR).toString()
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: See my previous comment about variable naming.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 14]
```diff
@@ -125,4 +126,12 @@ class KNoTMessageParser {
 
         return KNoTMessageRegistered(id, token, error)
     }
+
+    fun parseDeviceRemoved(knotRemove : String) : KNoTMessageRemoved {
+        val jelement : JsonElement = JsonParser.parseString(knotRemove)
+        val id = jelement.asJsonObject.get(ID).asString
+        val error = jelement.asJsonObject.get(ERROR).toString()
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: See my previous comment about variable naming.

**Review by lcc — CHANGES_REQUESTED** [review]

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt (hunk 5 above)
Comment (alla3) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt:27]: It is described as 'sensorId' on the KNoT Fog Connector API.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt (hunk 7 above)
Comment (alla3) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: I agree.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageAuthenticated.kt (hunk 3 above)
Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageAuthenticated.kt]: Could you please address this issue?

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 15]
```diff
@@ -92,4 +94,12 @@ class KNoTMessageParser {
 
         return KNoTMessageRequestData(id, sensorIds)
     }
+
+    fun parseSchemaStatus(knotSchema : String) : KNoTMessageSchemaResp {
+        val receivedJson : JsonElement = JsonParser.parseString(knotSchema)
+        val id = receivedJson.asJsonObject.get(ID).asString
+        val error = receivedJson.asJsonObject.get(ERROR).toString()
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: I believe you could change the first line to end with asJsonObject instead of calling it twice down bellow. Look at your previous commit.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 16]
```diff
@@ -102,4 +103,12 @@ class KNoTMessageParser {
 
         return KNoTMessageSchemaResp(id, error)
     }
+
+    fun parseAuthStatus(knotAuth : String) : KNoTMessageAuthenticated {
+        val receivedJson : JsonElement = JsonParser.parseString(knotAuth)
+        val id = receivedJson.asJsonObject.get(ID).asString
+        val error = receivedJson.asJsonObject.get(ERROR).toString()
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: Look at my previous commit.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 17]
```diff
@@ -111,4 +112,13 @@ class KNoTMessageParser {
 
         return KNoTMessageAuthenticated(id, error)
     }
+
+    fun parseDeviceRegistered(knotRegister : String) : KNoTMessageRegistered {
+        val receivedJson : JsonElement = JsonParser.parseString(knotRegister)
+        val id = receivedJson.asJsonObject.get(ID).asString
+        val token = receivedJson.asJsonObject.get(TOKEN).toString()
+        val error = receivedJson.asJsonObject.get(ERROR).toString()
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: Look at my previous commit.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt [hunk 18]
```diff
@@ -121,4 +122,12 @@ class KNoTMessageParser {
 
         return KNoTMessageRegistered(id, token, error)
     }
+
+    fun parseDeviceRemoved(knotRemove : String) : KNoTMessageRemoved {
+        val receivedJson : JsonElement = JsonParser.parseString(knotRemove)
+        val id = receivedJson.asJsonObject.get(ID).asString
+        val error = receivedJson.asJsonObject.get(ERROR).toString()
```
Comment (lcc) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: Look at my previous commit.

**Review by lcc — CHANGES_REQUESTED** [review]

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt (hunk 15 above)
Comment (alla3) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: Done.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageAuthenticated.kt (hunk 3 above)
Comment (alla3) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageAuthenticated.kt]: Done.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt (hunk 16 above)
Comment (alla3) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: Done.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageRegistered.kt (hunk 4 above)
Comment (alla3) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageRegistered.kt:12]: The API has an error but the example is correct.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt (hunk 17 above)
Comment (alla3) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: Done.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt (hunk 18 above)
Comment (alla3) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessageParser.kt]: Done.

**Review by lcc — APPROVED** [review]
LGTM
